src/upgrade.rs

```diff
@@ -0,0 +1,152 @@
+//! Everything related to the upgrade process.
+
+use heed::{
+    types::{Bytes, LazyDecode, Unit},
+    RoTxn, RwTxn,
+};
+
+use crate::{
+    distance::Cosine,
+    key::Key,
+    metadata::MetadataCodec,
+    node::{Node, NodeCodec},
+    node_id::NodeMode,
+    roaring::RoaringBitmapCodec,
+    version::{Version, VersionCodec},
+    Database, Distance, Error, Result,
+};
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(u8)]
+enum OldNodeMode {
```

Comment:
```suggestion
enum NodeMode_0_4 {
```

Since it may change again maybe we should specify the version what do you think? 🤔